README.md
# pygeems
Geotechnical earthquake engineering models implemented in Python.


# Code style

Docstrings should conform to the numpy documentation style: https://numpydoc.readthedocs.io/en/latest/format.html


HISTORY.rst
History
=======

v0.2.0 (2020-03-02)
-------------------
- Added Bray slope models

v0.1.0
------
- Initial release


tests/__init__.py
import pathlib

FPATH_DATA = pathlib.Path(__file__).parent / "data"


tests/test_slope_disp.py
import pathlib

import numpy as np
import pytest
from numpy.testing import assert_allclose

import pygeems


FPATH_DATA = pathlib.Path(__file__).parent / "data"


@pytest.mark.parametrize(
    "yield_coef,pga,mag,pgv,expected",
    [
        # Examples from Rathje et. al (2014)
        (0.1, 0.54, 6.75, None, 43),
        (0.1, 0.54, None, 42, 29),
        (0.1, 0.88, 6.75, None, 113),
        (0.1, 0.88, None, 71, 81),
        # Examples from Rathje & Antonakos (2011)
        # These do not work...
        # (0.1, 0.38, 8.0, None, 63.1),
        # (0.1, 0.38, None, 80, 36.9),
    ],
)
def test_calc_disp_rs08(yield_coef, pga, mag, pgv, expected):
    disp = pygeems.slope_disp.calc_disp_rs08(yield_coef, pga, mag=mag, pgv=pgv)
    # Compare to 2 cm of displacement
    assert_allclose(disp, expected, atol=2)


@pytest.mark.parametrize(
    "yield_coef,pga,period_slide,period_mean,mag,pgv,expected",
    [
        # Examples from Rathje et. al (2014)
        # Values read from Figure 8c
        (0.05, 0.35, 0.2, 0.45, 7.0, None, 87.5),
        (0.10, 0.35, 0.2, 0.45, 7.0, None, 22.0),
        (0.05, 0.35, 0.2, 0.45, None, 30, 43.5),
        (0.10, 0.35, 0.2, 0.45, None, 30, 10.5),
        # Examples from Rathje & Antonakos (2011)
        # These do not work...
        # (0.1, 0.48, 0.2, 0.46, 8.0, None, 126),
        # (0.1, 0.48, 0.2, 0.46, None, 74, 49)
    ],
)
def test_calc_disp_ra11(yield_coef, pga, period_slide, period_mean, mag, pgv, expected):
    disp = pygeems.slope_disp.calc_disp_ra11(
        yield_coef, pga, period_slide, period_mean, mag=mag, pgv=pgv
    )
    assert_allclose(disp, expected, atol=1)


def test_calc_wla06_ln_a_rms():
    a_rms = np.exp(pygeems.slope_disp._calc_wla06_ln_a_rms(0.35))
    assert_allclose(a_rms, 0.1067, atol=0.0001)


def test_calc_wla06_ln_dur_key():
    dur_ky = np.exp(pygeems.slope_disp._calc_wla06_ln_dur_key(0.11, 0.35, 0.18, 6.5))
    assert_allclose(dur_ky, 0.5896, atol=0.0001)


def test_calc_disp_wla06():
    median = pygeems.slope_disp.calc_disp_wla06(0.11, 0.35, 0.18, 6.5)
    assert_allclose(median, 3.001, atol=0.001)


# Test values from Table 1 in BT07
@pytest.mark.parametrize(
    "yield_coef,period_slide,psa_dts,expected",
    [
        (0.35, 0.45, 0.43, 0.85),
        (0.08, 0.00, 0.24, 0.1),
        (0.14, 0.33, 0.94, 0.0),
    ],
)
def test_calc_prob_disp_bt07(yield_coef, period_slide, psa_dts, expected):
    # Convert from probability of zero displacement to probability of non-zero
    # displacement
    expected = 1 - expected

    calc = pygeems.slope_disp.calc_prob_disp_bt07(yield_coef, period_slide, psa_dts)
    assert_allclose(calc, expected, atol=0.02, rtol=0.02)


# Test values from Table 1 in BT07. Authors provided a range of values, which are
# interpreted to be log-normally distributed. The median is computed and tested against
@pytest.mark.parametrize(
    "yield_coef,period_slide,psa_dts,expected",
    [
        (0.08, 0.00, 0.24, np.sqrt(4 * 15)),
        (0.14, 0.33, 0.94, np.sqrt(20 * 70)),
    ],
)
def test_calc_disp_bt07(yield_coef, period_slide, psa_dts, expected):
    median = pygeems.slope_disp.calc_disp_bt07(yield_coef, period_slide, psa_dts)
    assert_allclose(median, expected, rtol=0.10)


@pytest.mark.parametrize(
    "invert,yield_coef,expected",
    [
        (False, 0.05, 69.4),
        (True, 0.05, 66.1),
        (False, 0.10, 28.5),
        (True, 0.10, 30.5),
    ],
)
def test_calc_block_displacement(invert, yield_coef, expected):
    time_step = 0.01
    # Acceleration data in g
    accels = np.loadtxt(str(FPATH_DATA / "kobe-nis-000.dat"))
    yield_coef = yield_coef

    disps, vels = pygeems.slope_disp.calc_rigid_disp(
        time_step, accels, yield_coef, invert
    )
    disp_max = disps[-1]
    # Large rtol because of slightly different implementations
    assert_allclose(disp_max, expected, rtol=0.04)


# Test values from Table 2 of Bray et al. (2018). Coastline slope and Nishigo dam
@pytest.mark.parametrize(
    "yield_coef,period_slide,psa_dts,mag,expected",
    [
        (0.10, 0.6, 0.25, 8.0, np.sqrt(3 * 12)),
        (0.26, 0.15, 1.51, 9.0, np.sqrt(14 * 58)),
    ],
)
def test_calc_disp_bea18(yield_coef, period_slide, psa_dts, mag, expected):
    ret = pygeems.slope_disp.calc_disp_bea18(
        yield_coef, period_slide, psa_dts, mag, stats=False
    )
    # Limited number of digits in table make testing difficult
    assert_allclose(ret, expected, rtol=0.50)


# Test values from Table 2 of Bray et al. (2018). Esperanza and Tutuven dams.
@pytest.mark.parametrize(
    "yield_coef,period_slide,psa_dts,expected",
    [
        (0.24, 0.40, 0.43, 0.50),
        (0.39, 0.15, 0.75, 0.60),
    ],
)
def test_calc_prob_disp_bea18(yield_coef, period_slide, psa_dts, expected):
    ret = pygeems.slope_disp.calc_prob_disp_bea18(yield_coef, period_slide, psa_dts)
    # Reported values are probability of no displacement
    # Limited number of digits in table make testing difficult
    assert_allclose(ret, 1 - expected, rtol=0.25)


pygeems/utils.py
import functools

import numpy as np


def check_bounds(value, lower, upper, name, format="f"):
    if lower and np.any(value < lower):
        raise ValueError(
            f'"{name}" value of {value:{format}} below lower limit of {lower:{format}}'
        )

    if upper and np.any(value > upper):
        raise ValueError(
            f'"{name}" value of {value:{format}} above upper limit of {upper:{format}}'
        )


def check_options(value, options, name):
    if value not in options:
        raise ValueError(f'"{name}" not in options: {options}')


def dist_lognorm(f):
    @functools.wraps(f)
    def wrapper(*args, **kwds):
        ln_mean, ln_std = f(*args, **kwds)

        if kwds.get("stats", False):
            return ln_mean, ln_std
        else:
            return np.exp(ln_mean)

    return wrapper


pygeems/dyn_props.py
import numpy as np

from . import PRESS_ATM


def calc_mod_shear_ws18(
    stress_mean_eff,
    fines,
    plas_index=None,
    over_consol_ratio=1.0,
    unif_coef=None,
    diam_mean=None,
    water_content=None,
    void_ratio=None,
):
    """Compute the shear modulus based on Wang & Stokoe model."""

    if fines <= 0.12:
        # Clean sands and gravels
        c_g = 63.9e3
        f_g = unif_coef ** -0.21 * void_ratio ** (-1.12 - (0.09 * diam_mean) ** 0.54)
        stress_exp = 0.48 * unif_coef ** 0.08 - 1.03 * fines
    elif fines > 0.12 and plas_index is None:
        c_g = 84.8e3
        f_g = np.exp(-0.53) * (1 - 1.32 * water_content)
        stress_exp = 0.52
    elif fines > 0.12 and plas_index > 0:
        c_g = 232.9e3
        f_g = (
            (1 + 0.96 * void_ratio) ** -2.42
            * (1.92 + over_consol_ratio) ** (0.27 + 0.46 * plas_index)
            * (1 - 0.44 * fines)
        )
        stress_exp = 0.49
    else:
        raise NotImplementedError

    return c_g * f_g * stress_mean_eff ** stress_exp


def calc_vel_shear_spt_wds12(blows, stress_vert_eff, soil_type=all, age=None):
    """Estimate the shear-wave velocity based on blow count.

    Equations from Section 4.5 of Wair, Dejong, and Shantz (2012).

    Parameters
    ----------
    blows: `array_like`
        Blow counts, N_60
    stress_vert_eff: `array_like`
        Vertical effective stress (atm)
    soil_type: str, optional
        Soil type. Possible options include: 'fine_grained', 'sand', 'gravels',
        or 'all' (default).
    age: str, optional
        Age of soil. Possible options inlude: 'holocene' or 'pleistocene'. If
        *None*, then no scaling applied.
    Returns
    -------
    vel_shear: :class:`numpy.ndarray`
        Shear-wave velocity (m/s)
    """
    C = {
        "fine_grained": {
            "coeffs": (26, 0.17, 0.32),
            "asf": {
                "holocene": 0.88,
                "pleistocene": 1.12,
            },
        },
        "sand": {
            "coeffs": (30, 0.23, 0.23),
            "asf": {
                "holocene": 0.90,
                "pleistocene": 1.17,
            },
        },
        "gravel": {
            "coeffs": (78, 0.19, 0.18),
            "asf": {
                "holocene": 53 / 78,
                "pleistocene": 115 / 78,
            },
        },
        "all": {
            "coeffs": (30, 0.215, 0.275),
            "asf": {
                "holocene": 0.87,
                "pleistocene": 1.13,
            },
        },
    }

    # Convert to kPa
    stress_vert_eff = stress_vert_eff * PRESS_ATM

    coeff, pow_blows, pow_stress = C[soil_type]["coeffs"]
    vel_shear = coeff * blows ** pow_blows * stress_vert_eff ** pow_stress

    if age is not None:
        asf = C[soil_type]["asf"][age]
        vel_shear *= asf

    return vel_shear


pygeems/slope_disp.py
"""pyGEEMs: slope displacement functions and classes."""
import collections
from typing import Optional
from typing import Union

import numba
import numpy as np
import scipy.constants
from numpy.typing import ArrayLike
from scipy.integrate import cumtrapz
from scipy.stats import norm

from .utils import dist_lognorm

FloatOrArrayLike = Union[float, ArrayLike]


@dist_lognorm
def calc_disp_rs08(
    yield_coef: float,
    pga: FloatOrArrayLike,
    mag: Optional[float] = None,
    pgv: Optional[float] = None,
    **kwargs
):
    """Rathje and Saygili (2008) slope displacement model.

    Parameters
    ----------
    yield_coef : float
        Slope yield coefficient
    pga : float or `array_like`
        Peak ground acceleration in [gravity]
    mag : float
        Earthquake magnitude
    pgv : float
        Peak ground velocity in [cm/s]
    stats : bool, default=False
        If a distribution should be returned

    Returns
    -------
    median : float or `numpy.ndarray`
        Median value. If `stats` is False.
    ln_mean : float or `numpy.ndarray`
        Natural log mean value. Returned if `stats` is True.
    ln_std : float or `numpy.ndarray`
        Natural log standard deviation. Returned if `stats` is True.
    """
    if mag is None:
        method = "pgv"
    elif pgv is None:
        method = "mag"
    else:
        raise NotImplementedError

    P = {
        model: np.rec.fromrecords(values, names="a1,a2,a3,a4,a5,a6,a7")
        for model, values in zip(
            ["mag", "pgv"],
            [
                (4.89, -4.85, -19.64, 42.49, -29.06, 0.72, 0.89),
                (-1.56, -4.58, -20.84, 44.75, -30.50, -0.64, 1.55),
            ],
        )
    }[method]

    yp_ratio = yield_coef / pga
    ln_disp = (
        P.a1
        + P.a2 * yp_ratio
        + P.a3 * yp_ratio ** 2
        + P.a4 * yp_ratio ** 3
        + P.a5 * yp_ratio ** 4
        + P.a6 * np.log(pga)
    )
    if method == "mag":
        ln_disp += P.a7 * (mag - 6)
    else:
        ln_disp += P.a7 * np.log(pgv)

    if method == "mag":
        ln_std = 0.73 + 0.79 * yp_ratio - 0.54 * yp_ratio ** 2
    else:
        ln_std = 0.41 + 0.52 * yp_ratio

    return ln_disp, ln_std


@dist_lognorm
def calc_disp_ra11(
    yield_coef: float,
    pga: FloatOrArrayLike,
    period_slide: float,
    period_mean: float,
    mag: Optional[float] = None,
    pgv: Optional[float] = None,
    **kwargs
):
    """Rathje and Antonakos (2011) slope displacement model."""
    if mag is None:
        method = "pgv"
    elif pgv is None:
        method = "mag"
    else:
        raise NotImplementedError

    period_ratio = period_slide / period_mean
    if period_ratio < 0.1:
        k_max = pga
    else:
        ln_pr = np.log(period_ratio / 0.1)
        ln_ratio = (0.459 - 0.702 * pga) * ln_pr + (-0.228 + 0.076 * pga) * ln_pr ** 2
        k_max = pga * np.exp(ln_ratio)

    # Compute median
    if method == "mag":
        ln_disp = calc_disp_rs08(yield_coef, k_max, mag=mag, stats=True)[0]
        f_1 = np.minimum(3.69 * period_slide - 1.22 * period_slide ** 2, 2.78)
        ln_disp += f_1
    else:
        if period_ratio < 0.2:
            k_velmax = pgv
        else:
            ln_pr = np.log(period_ratio / 0.2)
            ln_ratio = 0.240 * ln_pr + (-0.091 - 0.171 * pga) * ln_pr ** 2
            k_velmax = pgv * np.exp(ln_ratio)

        ln_disp = calc_disp_rs08(yield_coef, k_max, pgv=k_velmax, stats=True)[0]

        f_2 = np.minimum(1.42 * period_slide, 0.71)
        ln_disp += f_2

    # Compute the standard deviation
    if method == "mag":
        ln_std = 0.694 + 0.322 * (yield_coef / k_max)
    else:
        ln_std = 0.400 + 0.284 * (yield_coef / k_max)

    return ln_disp, ln_std


def _calc_wla06_ln_a_rms(pga):
    """Calculate $a_{rms}$ from WLA06.

    Provided by Equation (2).
    """
    return -1.167 + 1.02 * np.log(pga)


def _calc_wla06_ln_dur_key(yield_coef: float, pga: float, psa_1s: float, mag: float):
    """Watson-Lamprey and Abrahamson (2006) duration calculation.

    FIXME

    """
    # Simplification
    ln_pga = np.log(pga)
    ln_pga_ky = np.log(pga / yield_coef)
    ln_psa_1s = np.log(psa_1s)

    return (
        -2.775
        + 0.956 * ln_pga_ky
        - 1.554 / (ln_pga_ky + 0.390)
        - 0.597 * ln_pga
        + 0.381 * ln_psa_1s
        + 0.334 * mag
    )


@dist_lognorm
def calc_disp_wla06(yield_coef: float, pga: float, psa_1s: float, mag: float, **kwds):
    """Watson-Lamprey and Abrahamson (2006) slope displacement model."""
    # Constants from Table 1.
    # Updated values were provided by Jennie
    a1 = 5.463
    b1 = 0.451
    b2 = 0.0191
    c1 = 0.591
    d1 = 0.205
    d2 = 0.0892
    e1 = 1.039
    e2 = 0.0427
    f1 = -1.409
    f2 = 0.1

    # Simplification
    ln_psa_1s = np.log(psa_1s)

    ln_dur_ky = _calc_wla06_ln_dur_key(yield_coef, pga, psa_1s, mag)
    ln_a_rms = _calc_wla06_ln_a_rms(pga)

    pga = np.asarray(pga)
    ln_pga_ky = np.log(pga / yield_coef)
    ln_psa_pga = np.log(psa_1s / pga)

    ln_mean = (
        a1
        + b1 * (ln_psa_1s + 0.45)
        + b2 * (ln_psa_1s + 0.45) ** 2
        + c1 * (ln_a_rms - 1)
        + d1 * ln_psa_pga
        + d2 * ln_psa_pga ** 2
        + e1 * (ln_dur_ky - 0.74)
        + e2 * (ln_dur_ky - 0.74) ** 2
        + 1 / (f1 * (ln_pga_ky + f2))
    )
    # Equation not valid for PGA values less than the yield coef
    try:
        ln_mean[pga < yield_coef] = np.nan
    except TypeError:
        if ln_mean < yield_coef:
            ln_mean = np.nan

    # From Figure 5
    ln_std = 0.53

    return ln_mean, ln_std


@dist_lognorm
def calc_disp_bt07(yield_coef: float, period_slide: float, psa_dts: float, **kwds):
    """Bray and Travasarou (2007)."""
    # Simplification
    ln_yield_coef = np.log(yield_coef)
    ln_psa_dts = np.log(psa_dts)

    # Median displacement
    # Equation (4)
    ln_mean = -0.22 if period_slide < 0.05 else -1.10
    ln_mean += (
        -2.83 * ln_yield_coef
        - 0.333 * ln_yield_coef ** 2
        + 0.566 * ln_yield_coef * ln_psa_dts
        + 3.04 * ln_psa_dts
        - 0.244 * ln_psa_dts ** 2
        + 1.5 * period_slide
    )
    ln_std = 0.66
    return ln_mean, ln_std


def calc_prob_disp_bt07(yield_coef: float, period_slide: float, psa_dts: float, **kwds):
    """Bray and Travasarou (2007), probability of non-zero displacement."""
    # Probability of a non-zero displacement
    # Modified from Equation (3)
    ln_yield_coef = np.log(yield_coef)

    prob_disp = norm.cdf(
        -1.76
        - 3.22 * ln_yield_coef
        - 0.484 * period_slide * ln_yield_coef
        + 3.52 * np.log(psa_dts)
    )

    return prob_disp


@dist_lognorm
def calc_disp_bea18(
    yield_coef: float, period_slide: float, psa_dts: float, mag: float, **kwds
):
    """Bray et al. (2018)."""
    ln_yield_coef = np.log(yield_coef)
    ln_psa_dts = np.log(psa_dts)

    ln_mean = (
        -3.535 * ln_yield_coef
        - 0.390 * ln_yield_coef ** 2
        + 0.538 * ln_yield_coef * ln_psa_dts
        + 3.060 * ln_psa_dts
        - 0.225 * ln_psa_dts ** 2
        + 0.550 * mag
    )
    if np.isclose(period_slide, 0.0):
        ln_mean -= 5.864
    else:
        if period_slide < 0.1:
            a1, a2, a3 = -5.864, -9.421, 0
        else:
            a1, a2, a3 = -6.896, 3.081, -0.803
        ln_mean += a1 + a2 * period_slide + a3 * period_slide ** 2

    ln_std = 0.73
    return ln_mean, ln_std


def calc_prob_disp_bea18(
    yield_coef: float, period_slide: float, psa_dts: float, **kwds
):
    """Bray et al. (2018) slope displacement model for subduction events."""
    ln_yield_coef = np.log(yield_coef)

    if period_slide <= 0.7:
        z = (
            -2.64
            - 3.20 * ln_yield_coef
            - 0.17 * ln_yield_coef ** 2
            - 0.49 * period_slide * ln_yield_coef
            + 2.09 * period_slide
            + 2.91 * np.log(psa_dts)
        )
    else:
        z = (
            -3.53
            - 4.78 * ln_yield_coef
            - 0.34 * ln_yield_coef ** 2
            - 0.30 * period_slide * ln_yield_coef
            - 0.67 * period_slide
            + 2.66 * np.log(psa_dts)
        )

    prob_disp = norm.cdf(z)
    return prob_disp


@numba.jit
def _calc_block_velocity(time_step: float, accels: ArrayLike, yield_coef: float):
    """Compute the velocity of a sliding block.

    The calculation is adapted from Slammer's source code[1]_. However,
    the displacement is computed outside of the numerical integration.

    .. [1] https://github.com/mjibson/slammer/blob/master/programs/slammer/analysis/Decoupled.java#L135

    Parameters
    ----------
    time_step : float
        time series step size [seconds]
    accels : array_like
        acceleration time series [cm/sec/sec]
    yield_coef : float
        yield coefficient of the slope [cm/sec/sec]

    Returns
    -------
    vels : :class:`numpy.array`
        velocity time series of the sliding mass
    """
    vels = np.zeros_like(accels)
    sliding = False

    for i in range(1, len(accels)):
        # Compute the velocity and displacement of the sliding mass
        accel_avg = 0.5 * (accels[i] + accels[i - 1])

        if sliding:
            vels[i] = vels[i - 1] + (accel_avg - yield_coef) * time_step
        else:
            vels[i] = 0

        # Check if sliding
        if sliding:
            if vels[i] <= 0:
                sliding = False
                vels[i] = 0
        else:
            if accels[i] > yield_coef:
                sliding = True

    return vels


def calc_rigid_disp(
    time_step: float, accels: ArrayLike, yield_coef: float, invert: bool = False
):
    """Compute the displacement and velocity of a rigid sliding mass.

    Parameters
    ----------
    time_step : float
        time series step size [seconds]
    accels : array_like
        acceleration time series [g]
    yield_coef : float
        yield coefficient of the slope [g]
    invert : bool (optional)
        if the time series should be inverted
    Returns
    -------
    disps : :class:`numpy.array`
        displacement time series of the sliding mass
    vels : :class:`numpy.array`
        velocity time series of the sliding mass
    """
    # Convert into cm/s/s
    scale = scipy.constants.g / scipy.constants.centi

    yield_coef *= scale
    # Only invert the acceleration time series
    if invert:
        scale *= -1
    # Don't update the array inplace
    accels = accels * scale

    vels = _calc_block_velocity(time_step, accels, yield_coef)
    disps = np.r_[0, cumtrapz(vels, dx=time_step)]

    return disps, vels


pygeems/ground_motion.py
"""Functions and classes for calculation of ground motion parameters."""
from typing import Optional

import numpy as np
from numpy.typing import ArrayLike

from . import FPATH_DATA
from .utils import check_bounds
from .utils import check_options
from .utils import dist_lognorm

_CACHE_REA15 = {}


def calc_damping_scaling_rea15(
    damping: float,
    mag: float,
    dist_rup: float,
    comp: str = "rotd50",
    periods: Optional[ArrayLike] = None,
) -> ArrayLike:
    """Compute the damping scaling proposed by Rezaeian et al. (2014).

    Parameters
    ----------
    damping: damping ratio of the oscillator in percent (2 for 2%)
    mag: earthquake magnitude
    dist_rup: closest distance to the rupture in [km]
    comp: component, can be either: 'rotd50', 'roti50', 'vertical'
    periods: *optional* periods to provide values.

    Returns
    -------
    ret : np.rec.array
        Array with columns period, damping scaling factor, and natural logarithmic standard deviation.
    """
    check_bounds(damping, 0.5, 30, "damping")
    check_options(comp, ["rotd50", "roti50", "vertical"], "comp")

    # Only load the coefficients once
    if comp not in _CACHE_REA15:
        _CACHE_REA15[comp] = np.genfromtxt(
            FPATH_DATA / f"rezaeian_et_al_2014-{comp}.csv",
            delimiter=",",
            skip_header=1,
            names=True,
        ).view(np.recarray)
    C = _CACHE_REA15[comp]

    ln_damp = np.log(damping)
    ln_dsf = (
        C.b0
        + C.b1 * ln_damp
        + C.b2 * ln_damp ** 2
        + (C.b3 + C.b4 * ln_damp + C.b5 * ln_damp ** 2) * mag
        + (C.b6 + C.b7 * ln_damp + C.b8 * ln_damp ** 2) * np.log(dist_rup + 1)
    )
    ln_damp_5 = np.log(damping / 5)
    ln_std = np.abs(C.a0 * ln_damp_5 + C.a1 * ln_damp_5 ** 2)

    if periods:
        # Interpolate over the provided periods
        ln_dsf = np.interp(np.log(periods), np.log(C.period), ln_dsf)
        ln_std = np.interp(np.log(periods), np.log(C.period), ln_std)
    else:
        periods = C.period

    ret = np.rec.fromarrays(
        [periods, np.exp(ln_dsf), ln_std], names="period,dsf,ln_std"
    )
    return ret


@dist_lognorm
def calc_period_rea05(
    kind: str,
    mag: float,
    dist_rup: float,
    site_class: str = "c",
    directivity: bool = False,
    **kwargs,
):
    """Rathje et al. (2005) period metrics."""
    # Model coefficients from Table 2
    C = {
        model: np.rec.fromrecords(values, names="c1,c2,c3,c4,c5,c6")
        for model, values in zip(
            ["period_mean", "period_avg", "period_pred"],
            [
                (-1.00, 0.18, 0.0038, 0.078, 0.27, 0.40),
                (-0.89, 0.29, 0.0030, 0.070, 0.25, 0.37),
                (-1.78, 0.30, 0.0045, 0.150, 0.33, 0.24),
            ],
        )
    }[kind]

    check_options(site_class, "bcd", "site_class")
    if kind in ["period_avg", "period_pred"]:
        check_bounds(mag, 4.7, 7.6, "mag")
    else:
        mag = np.minimum(mag, 7.25)
        check_bounds(mag, 5.0, 7.25, "mag")

    if site_class == "c":
        s_c = 1
        s_d = 0
    elif site_class == "d":
        s_c = 0
        s_d = 1
    else:
        s_c = 0
        s_d = 0

    f_d = int(directivity)

    ln_period = (
        C.c1
        + C.c2 * (mag - 6)
        + C.c3 * dist_rup
        + C.c4 * s_c
        + C.c5 * s_d
        + C.c6 * (1 - dist_rup / 20) * f_d
    )

    intra = {
        "period_mean": {"b": 0.42, "c": 0.38, "d": 0.31},
        "period_avg": {"b": 0.42, "c": 0.38, "d": 0.31},
        "period_pred": {"b": 0.42, "c": 0.38, "d": 0.31},
    }[kind][site_class]
    inter = {"period_mean": 0.17, "period_avg": 0.13, "period_pred": 0.22}[kind]
    ln_std = np.sqrt(intra ** 2 + inter ** 2)

    return ln_period, ln_std


@dist_lognorm
def calc_aris_intensity_aea16(
    mag: float,
    v_s30: float,
    pga: float,
    psa_1s: float,
    hanging_wall: bool = False,
    dist_jb: Optional[float] = None,
    ln_std_pga: Optional[float] = None,
    ln_std_psa_1s: Optional[float] = None,
    **kwargs,
):
    """Abrahamson et al. (2016) Arias intensity model."""
    # From Table 3.1
    # Value of c8 is provided after equation 3.7
    C = np.rec.fromrecords(
        (0.47, -0.28, 0.50, 1.52, 0.21, 0.09), names="c1,c2,c3,c4,c5,c8"
    )

    ln_arias_int = (
        C.c1
        + C.c2 * np.log(v_s30)
        + C.c3 * mag
        + C.c4 * np.log(pga)
        + C.c5 * np.log(psa_1s)
    )
    if hanging_wall:
        ln_arias_int += C.c8 * np.clip(1 - (dist_jb - 5) / 5, 0, 1)

    if ln_std_pga is None and ln_std_psa_1s is None:
        # Computed from residuals
        ln_std = np.interp(
            mag, [3, 4, 5, 6, 7], [0.40, 0.39, 0.37, 0.33, 0.40], left=0.40, right=0.40
        )
    else:
        raise NotImplementedError

    return ln_arias_int, ln_std
